File: backend/src/main/java/cn/etarch/mao/harness/mcp/service/McpServerService.java
package cn.etarch.mao.harness.mcp.service;

import cn.etarch.mao.agent.entity.Agent;
import cn.etarch.mao.agent.mapper.AgentMapper;
import cn.etarch.mao.common.exception.BusinessException;
import cn.etarch.mao.common.result.ErrorCode;
import cn.etarch.mao.harness.mcp.crypto.McpSecretCipher;
import cn.etarch.mao.harness.mcp.entity.McpServer;
import cn.etarch.mao.harness.mcp.mapper.McpServerMapper;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * MCP 服务器配置管理服务。
 * 负责 CRUD、字段校验、环境变量加密存储、Agent 引用检查。
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class McpServerService {

    private static final Pattern NAME_PATTERN = Pattern.compile("^[a-z0-9_-]+$");

    private final McpServerMapper mcpServerMapper;
    private final AgentMapper agentMapper;
    private final McpSecretCipher secretCipher;
    private final ObjectMapper objectMapper;

    public List<McpServer> list(String keyword, String status) {
        LambdaQueryWrapper<McpServer> qw = new LambdaQueryWrapper<>();
        if (StringUtils.hasText(keyword)) {
            qw.and(w -> w.like(McpServer::getName, keyword)
                    .or().like(McpServer::getDescription, keyword));
        }
        if (StringUtils.hasText(status)) {
            qw.eq(McpServer::getStatus, status);
        }
        qw.orderByAsc(McpServer::getId);
        List<McpServer> servers = mcpServerMapper.selectList(qw);
        // 列表不回传明文环境变量
        servers.forEach(s -> s.setEnvJson(null));
        return servers;
    }

    /** 仅返回已启用的服务器（供 Agent 关联勾选与运行时加载）。 */
    public List<McpServer> listEnabled() {
        LambdaQueryWrapper<McpServer> qw = new LambdaQueryWrapper<>();
        qw.eq(McpServer::getStatus, McpServer.STATUS_ENABLED);
        qw.orderByAsc(McpServer::getId);
        List<McpServer> servers = mcpServerMapper.selectList(qw);
        servers.forEach(s -> s.setEnvJson(null));
        return servers;
    }

    public McpServer get(Long id) {
        McpServer server = mcpServerMapper.selectById(id);
        if (server == null) {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "MCP 服务器不存在");
        }
        server.setEnvJson(null);
        return server;
    }

    /**
     * 运行时详情（含加密 env 原文，仅供服务端内部使用，如测试连接）。
     * 不经任何管理接口回传。
     */
    public McpServer getForRuntime(Long id) {
        McpServer server = mcpServerMapper.selectById(id);
        if (server == null) {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "MCP 服务器不存在");
        }
        return server;
    }

    public McpServer create(String name, String description, String serverType,
                            String command, List<String> args, String url,
                            Map<String, String> env) {
        validateName(name);
        McpServer server = new McpServer();
        server.setName(name);
        applyFields(server, description, serverType, command, args, url, env);
        server.setStatus(McpServer.STATUS_ENABLED);
        mcpServerMapper.insert(server);
        log.info("Created MCP server: id={}, name={}, type={}", server.getId(), name, serverType);
        return server;
    }

    public McpServer update(Long id, String name, String description, String serverType,
                            String command, List<String> args, String url,
                            Map<String, String> env) {
        McpServer server = mcpServerMapper.selectById(id);
        if (server == null) {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "MCP 服务器不存在");
        }
        // 名称变更时校验唯一性（排除自身）
        if (StringUtils.hasText(name) && !name.equals(server.getName())) {
            validateName(name);
        }
        if (StringUtils.hasText(name)) {
            server.setName(name);
        }
        applyFields(server, description, serverType, command, args, url, env);
        mcpServerMapper.updateById(server);
        log.info("Updated MCP server: id={}, name={}", id, server.getName());
        return server;
    }

    public void updateStatus(Long id, String status) {
        if (!McpServer.STATUS_ENABLED.equals(status) && !McpServer.STATUS_DISABLED.equals(status)) {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "状态值不合法");
        }
        McpServer server = mcpServerMapper.selectById(id);
        if (server == null) {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "MCP 服务器不存在");
        }
        server.setStatus(status);
        mcpServerMapper.updateById(server);
        log.info("Updated MCP server status: id={}, status={}", id, status);
    }

    public void delete(Long id) {
        McpServer server = mcpServerMapper.selectById(id);
        if (server == null) {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "MCP 服务器不存在");
        }
        List<Agent> referencingAgents = findReferencingAgents(id);
        if (!referencingAgents.isEmpty()) {
            List<String> names = referencingAgents.stream().map(Agent::getName).toList();
            throw new BusinessException(ErrorCode.PARAM_INVALID,
                    "该 MCP 服务器正被 Agent 引用（" + String.join("、", names) + "），请先解除关联");
        }
        mcpServerMapper.deleteById(id);
        log.info("Deleted MCP server: id={}, name={}", id, server.getName());
    }

    /**
     * 校验 Agent 关联的 MCP 服务器 ID 列表：去重、必须存在且已启用。
     * 任一 ID 无效时抛出 PARAM_INVALID（防止 Agent 配置引用不存在的服务器）。
     *
     * @return 规范化后的 ID 列表（保持原顺序、去重）
     */
    public List<Long> validateForAgent(List<Long> ids) {
        if (ids == null || ids.isEmpty()) {
            return List.of();
        }
        List<Long> result = new ArrayList<>();
        for (Long id : ids) {
            if (id == null) {
                continue;
            }
            if (result.contains(id)) {
                continue;
            }
            McpServer server = mcpServerMapper.selectById(id);
            if (server == null) {
                throw new BusinessException(ErrorCode.PARAM_INVALID, "MCP 服务器不存在（id=" + id + "）");
            }
            if (!McpServer.STATUS_ENABLED.equals(server.getStatus())) {
                throw new BusinessException(ErrorCode.PARAM_INVALID,
                        "MCP 服务器「" + server.getName() + "」已停用，无法关联");
            }
            result.add(id);
        }
        return result;
    }

    /**
     * 查询引用了指定 MCP 服务器的 Agent 列表（mcp_server_ids JSON 中包含该 ID）。
     */
    public List<Agent> findReferencingAgents(Long mcpServerId) {
        List<Agent> all = agentMapper.selectList(new LambdaQueryWrapper<Agent>());
        List<Agent> referencing = new ArrayList<>();
        for (Agent agent : all) {
            if (agent.getMcpServerIds() == null || agent.getMcpServerIds().isBlank()) {
                continue;
            }
            try {
                List<Long> ids = objectMapper.readValue(
                        agent.getMcpServerIds(), new TypeReference<List<Long>>() {});
                if (ids != null && ids.contains(mcpServerId)) {
                    referencing.add(agent);
                }
            } catch (Exception e) {
                log.warn("Failed to parse mcpServerIds for agent {}: {}", agent.getId(), e.getMessage());
            }
        }
        return referencing;
    }

    /**
     * 解密环境变量 JSON → Map。
     * 供运行时（CLOUD 直连 / LOCAL 下发）读取真实环境变量。
     */
    @SuppressWarnings("unchecked")
    public Map<String, String> decryptEnv(McpServer server) {
        if (server.getEnvJson() == null || server.getEnvJson().isBlank()) {
            return Map.of();
        }
        String plain = secretCipher.decrypt(server.getEnvJson());
        try {
            Map<String, String> env = objectMapper.readValue(plain,
                    new TypeReference<Map<String, String>>() {});
            return env != null ? env : Map.of();
        } catch (Exception e) {
            log.error("Failed to parse decrypted env for MCP server {}", server.getId(), e);
            throw new BusinessException(ErrorCode.INTERNAL_ERROR, "MCP 环境变量解析失败");
        }
    }

    private void applyFields(McpServer server, String description, String serverType,
                             String command, List<String> args, String url,
                             Map<String, String> env) {
        if (description != null) {
            server.setDescription(description);
        }
        if (serverType != null) {
            if (!McpServer.TYPE_STDIO.equals(serverType) && !McpServer.TYPE_HTTP.equals(serverType)) {
                throw new BusinessException(ErrorCode.PARAM_INVALID, "服务器类型必须为 STDIO 或 HTTP");
            }
            server.setServerType(serverType);
        }
        if (command != null) {
            server.setCommand(command);
        }
        if (args != null) {
            try {
                server.setArgsJson(objectMapper.writeValueAsString(args));
            } catch (Exception e) {
                throw new BusinessException(ErrorCode.PARAM_INVALID, "启动参数格式错误");
            }
        }
        if (url != null) {
            server.setUrl(url);
        }
        if (env != null) {
            try {
                server.setEnvJson(secretCipher.encrypt(objectMapper.writeValueAsString(env)));
            } catch (Exception e) {
                throw new BusinessException(ErrorCode.PARAM_INVALID, "环境变量格式错误");
            }
        }
        validateRequiredFields(server);
    }

    private void validateName(String name) {
        if (!StringUtils.hasText(name)) {
            throw new BusinessException(ErrorCode.PARAM_MISSING, "名称不能为空");
        }
        if (!NAME_PATTERN.matcher(name).matches()) {
            throw new BusinessException(ErrorCode.PARAM_INVALID,
                    "名称只能包含小写字母、数字、下划线和连字符");
        }
        // 工具名格式为 mcp__{serverName}__{toolName}，桌面端以第一个 __ 拆分服务器名与工具名。
        // 若服务器名本身含连续下划线（如 foo__bar），解析将产生歧义导致 LOCAL 模式无法调用。
        if (name.contains("__")) {
            throw new BusinessException(ErrorCode.PARAM_INVALID,
                    "名称不能包含连续下划线（__），请使用单个下划线或连字符分隔");
        }
        Long count = mcpServerMapper.selectCount(
                new LambdaQueryWrapper<McpServer>().eq(McpServer::getName, name));
        if (count != null && count > 0) {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "MCP 服务器名称已存在");
        }
    }

    private void validateRequiredFields(McpServer server) {
        String type = server.getServerType();
        if (McpServer.TYPE_STDIO.equals(type)) {
            if (!StringUtils.hasText(server.getCommand())) {
                throw new BusinessException(ErrorCode.PARAM_MISSING, "STDIO 类型必须填写启动命令");
            }
            if (!StringUtils.hasText(server.getArgsJson()) || "[]".equals(server.getArgsJson())) {
                throw new BusinessException(ErrorCode.PARAM_MISSING, "STDIO 类型必须填写启动参数");
            }
        } else if (McpServer.TYPE_HTTP.equals(type)) {
            if (!StringUtils.hasText(server.getUrl())) {
                throw new BusinessException(ErrorCode.PARAM_MISSING, "HTTP 类型必须填写服务器 URL");
            }
            if (!server.getUrl().startsWith("http://") && !server.getUrl().startsWith("https://")) {
                throw new BusinessException(ErrorCode.PARAM_INVALID, "URL 必须以 http:// 或 https:// 开头");
            }
        } else {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "服务器类型必须为 STDIO 或 HTTP");
        }
    }
}

File: backend-ts/src/harness/mcp/service/mcp-server.service.ts
import { BusinessException } from '../../../common/business-exception.js';
import { ErrorCode } from '../../../common/error-code.js';
import { hasText } from '../../../common/case.js';
import { harnessLog } from '../../log.js';
import type { Agent } from '../../deps.js';
import type { McpSecretCipher } from '../crypto/mcp-secret-cipher.js';
import {
  GLOBAL_USER_ID, STATUS_DISABLED, STATUS_ENABLED, TYPE_HTTP, TYPE_STDIO, type McpServer,
} from '../entity/mcp-server.js';
import type { McpServerMapper } from '../mapper/mcp-server.mapper.js';
import type { UserMcpPreferenceService } from '../preference/service/user-mcp-preference.service.js';

const NAME_PATTERN = /^[a-z0-9_-]+$/;

export { GLOBAL_USER_ID };

export interface UserLookup {
  findById(id: number): Promise<{ username?: string | null; displayName?: string | null } | null>;
}

export interface AgentListMapper {
  selectById(id: number): Promise<Agent | null>;
  listAll?(): Promise<Agent[]>;
  selectList?(): Promise<Agent[]>;
}

export class McpServerService {
  constructor(
    private readonly mapper: McpServerMapper,
    private readonly secretCipher: McpSecretCipher,
    private readonly preferenceService: UserMcpPreferenceService,
    private readonly userLookup?: UserLookup | null,
    private readonly agentMapper?: AgentListMapper | null,
  ) {}

  async list(keyword?: string | null, status?: string | null): Promise<McpServer[]> {
    const servers = await this.mapper.list(keyword, status);
    await this.fillOwnerNames(servers);
    for (const s of servers) s.envJson = null;
    return servers;
  }

  async listEnabled(): Promise<McpServer[]> {
    const servers = await this.mapper.listEnabledGlobal();
    for (const s of servers) s.envJson = null;
    return servers;
  }

  async listMine(userId: number | null | undefined): Promise<McpServer[]> {
    if (userId == null) return [];
    const servers = await this.mapper.listMine(userId);
    for (const s of servers) s.envJson = null;
    return servers;
  }

  async get(id: number): Promise<McpServer> {
    const server = await this.mapper.selectById(id);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器不存在');
    server.envJson = null;
    return server;
  }

  async getMine(userId: number, id: number): Promise<McpServer> {
    const server = await this.mapper.selectById(id);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器不存在');
    if (server.userId !== userId) throw new BusinessException(403, '无权操作该 MCP 服务器');
    server.envJson = null;
    return server;
  }

  async getForRuntime(id: number): Promise<McpServer> {
    const server = await this.mapper.selectById(id);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器不存在');
    return server;
  }

  async getMineForRuntime(userId: number, id: number): Promise<McpServer> {
    const server = await this.mapper.selectById(id);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器不存在');
    if (server.userId !== userId) throw new BusinessException(403, '无权操作该 MCP 服务器');
    return server;
  }

  create(
    name: string, description: string | null | undefined, serverType: string,
    command: string | null | undefined, args: string[] | null | undefined,
    url: string | null | undefined, env: Record<string, string> | null | undefined,
  ): Promise<McpServer> {
    return this.createInternal(GLOBAL_USER_ID, name, description, serverType, command, args, url, env);
  }

  createMine(
    userId: number, name: string, description: string | null | undefined, serverType: string,
    command: string | null | undefined, args: string[] | null | undefined,
    url: string | null | undefined, env: Record<string, string> | null | undefined,
  ): Promise<McpServer> {
    return this.createInternal(userId, name, description, serverType, command, args, url, env);
  }

  private async createInternal(
    ownerUserId: number, name: string, description: string | null | undefined, serverType: string,
    command: string | null | undefined, args: string[] | null | undefined,
    url: string | null | undefined, env: Record<string, string> | null | undefined,
  ): Promise<McpServer> {
    await this.validateName(name, ownerUserId);
    const server: McpServer = {
      userId: ownerUserId,
      name,
      status: STATUS_ENABLED,
    };
    this.applyFields(server, description, serverType, command, args, url, env);
    const id = await this.mapper.insert(server);
    server.id = id;
    harnessLog('info', `Created MCP server: id=${id}, name=${name}, type=${serverType}, owner=${ownerUserId}`);
    return server;
  }

  async update(
    id: number, name: string | null | undefined, description: string | null | undefined, serverType: string | null | undefined,
    command: string | null | undefined, args: string[] | null | undefined,
    url: string | null | undefined, env: Record<string, string> | null | undefined,
  ): Promise<McpServer> {
    const server = await this.mapper.selectById(id);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器不存在');
    if ((server.userId ?? 0) !== GLOBAL_USER_ID) {
      throw new BusinessException(403, '无权编辑用户私有 MCP 服务器');
    }
    return this.updateInternal(server, name, description, serverType, command, args, url, env);
  }

  async updateMine(
    userId: number, id: number, name: string | null | undefined, description: string | null | undefined,
    serverType: string | null | undefined, command: string | null | undefined,
    args: string[] | null | undefined, url: string | null | undefined,
    env: Record<string, string> | null | undefined,
  ): Promise<McpServer> {
    const server = await this.mapper.selectById(id);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器不存在');
    if (server.userId !== userId) throw new BusinessException(403, '无权操作该 MCP 服务器');
    return this.updateInternal(server, name, description, serverType, command, args, url, env);
  }

  private async updateInternal(
    server: McpServer, name: string | null | undefined, description: string | null | undefined,
    serverType: string | null | undefined, command: string | null | undefined,
    args: string[] | null | undefined, url: string | null | undefined,
    env: Record<string, string> | null | undefined,
  ): Promise<McpServer> {
    if (hasText(name) && name !== server.name) {
      await this.validateName(name!, server.userId ?? GLOBAL_USER_ID);
    }
    if (hasText(name)) server.name = name as string;
    this.applyFields(server, description, serverType, command, args, url, env);
    await this.mapper.updateById(server.id!, {
      name: server.name,
      description: server.description,
      serverType: server.serverType,
      command: server.command,
      argsJson: server.argsJson,
      url: server.url,
      envJson: server.envJson,
      status: server.status,
    });
    harnessLog('info', `Updated MCP server: id=${server.id}, name=${server.name}`);
    return server;
  }

  async updateStatus(id: number, status: string): Promise<void> {
    if (status !== STATUS_ENABLED && status !== STATUS_DISABLED) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '状态值不合法');
    }
    const server = await this.mapper.selectById(id);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器不存在');
    await this.mapper.updateById(id, { status });
    harnessLog('info', `Updated MCP server status: id=${id}, status=${status}`);
  }

  async delete(id: number): Promise<void> {
    const server = await this.mapper.selectById(id);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器不存在');
    if ((server.userId ?? 0) === GLOBAL_USER_ID) {
      const referencing = await this.findReferencingAgents(id);
      if (referencing.length > 0) {
        const names = referencing.map((a) => a.name).join('、');
        throw new BusinessException(ErrorCode.PARAM_INVALID, `该 MCP 服务器正被 Agent 引用（${names}），请先解除关联`);
      }
    }
    await this.mapper.physicalDeleteById(id);
    await this.preferenceService.deleteByServer(id);
    harnessLog('info', `Deleted MCP server: id=${id}, name=${server.name}`);
  }

  async deleteMine(userId: number, id: number): Promise<void> {
    const server = await this.mapper.selectById(id);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器不存在');
    if (server.userId !== userId) throw new BusinessException(403, '无权操作该 MCP 服务器');
    await this.mapper.physicalDeleteById(id);
    await this.preferenceService.deleteByServer(id);
    harnessLog('info', `Deleted user MCP server: id=${id}, name=${server.name}, owner=${userId}`);
  }

  async validateForAgent(ids: number[] | null | undefined): Promise<number[]> {
    if (!ids || ids.length === 0) return [];
    const result: number[] = [];
    for (const id of ids) {
      if (id == null || result.includes(id)) continue;
      const server = await this.mapper.selectById(id);
      if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, `MCP 服务器不存在（id=${id}）`);
      if ((server.userId ?? 0) !== GLOBAL_USER_ID) {
        throw new BusinessException(ErrorCode.PARAM_INVALID, `「${server.name}」为用户私有服务器，不能被 Agent 关联`);
      }
      if (server.status !== STATUS_ENABLED) {
        throw new BusinessException(ErrorCode.PARAM_INVALID, `MCP 服务器「${server.name}」已停用，无法关联`);
      }
      result.push(id);
    }
    return result;
  }

  async validatePreferenceTarget(userId: number, serverId: number): Promise<void> {
    const server = await this.mapper.selectById(serverId);
    if (!server) throw new BusinessException(ErrorCode.PARAM_INVALID, `MCP 服务器不存在（id=${serverId}）`);
    const isGlobal = (server.userId ?? 0) === GLOBAL_USER_ID;
    const isOwn = userId != null && server.userId === userId;
    if (!isGlobal && !isOwn) throw new BusinessException(403, '无权操作该 MCP 服务器');
    if (server.status !== STATUS_ENABLED) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, `MCP 服务器「${server.name}」已停用，无法启用`);
    }
  }

  async findReferencingAgents(mcpServerId: number): Promise<Agent[]> {
    const all = await (this.agentMapper?.listAll?.() ?? this.agentMapper?.selectList?.() ?? []);
    const referencing: Agent[] = [];
    for (const agent of all) {
      if (!agent.mcpServerIds || agent.mcpServerIds.trim() === '') continue;
      try {
        const ids = JSON.parse(agent.mcpServerIds) as number[];
        if (Array.isArray(ids) && ids.includes(mcpServerId)) referencing.push(agent);
      } catch (e) {
        harnessLog('warn', `Failed to parse mcpServerIds for agent ${agent.id}: ${(e as Error).message}`);
      }
    }
    return referencing;
  }

  decryptEnv(server: McpServer): Record<string, string> {
    if (!server.envJson || server.envJson.trim() === '') return {};
    const plain = this.secretCipher.decrypt(server.envJson);
    try {
      const env = JSON.parse(plain ?? '{}') as Record<string, string>;
      return env && typeof env === 'object' ? env : {};
    } catch (e) {
      harnessLog('error', `Failed to parse decrypted env for MCP server ${server.id}`, e);
      throw new BusinessException(ErrorCode.INTERNAL_ERROR, 'MCP 环境变量解析失败');
    }
  }

  private applyFields(
    server: McpServer, description: string | null | undefined, serverType: string | null | undefined,
    command: string | null | undefined, args: string[] | null | undefined,
    url: string | null | undefined, env: Record<string, string> | null | undefined,
  ): void {
    if (description !== undefined && description !== null) server.description = description;
    if (serverType != null) {
      if (serverType !== TYPE_STDIO && serverType !== TYPE_HTTP) {
        throw new BusinessException(ErrorCode.PARAM_INVALID, '服务器类型必须为 STDIO 或 HTTP');
      }
      server.serverType = serverType;
    }
    if (command !== undefined && command !== null) server.command = command;
    if (args != null) {
      try {
        server.argsJson = JSON.stringify(args);
      } catch {
        throw new BusinessException(ErrorCode.PARAM_INVALID, '启动参数格式错误');
      }
    }
    if (url !== undefined && url !== null) server.url = url;
    if (env != null) {
      try {
        server.envJson = this.secretCipher.encrypt(JSON.stringify(env)) ?? null;
      } catch (e) {
        if (e instanceof BusinessException) throw e;
        throw new BusinessException(ErrorCode.PARAM_INVALID, '环境变量格式错误');
      }
    }
    this.validateRequiredFields(server);
  }

  private async validateName(name: string, ownerUserId: number): Promise<void> {
    if (!hasText(name)) throw new BusinessException(ErrorCode.PARAM_MISSING, '名称不能为空');
    if (!NAME_PATTERN.test(name)) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '名称只能包含小写字母、数字、下划线和连字符');
    }
    if (name.includes('__')) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '名称不能包含连续下划线（__），请使用单个下划线或连字符分隔');
    }
    const count = await this.mapper.countByUserIdAndName(ownerUserId, name);
    if (count > 0) throw new BusinessException(ErrorCode.PARAM_INVALID, 'MCP 服务器名称已存在');
    const crossCount = ownerUserId === GLOBAL_USER_ID
      ? await this.mapper.countByNameWhereUserIdNot(name, GLOBAL_USER_ID)
      : await this.mapper.countByUserIdAndName(GLOBAL_USER_ID, name);
    if (crossCount > 0) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '该名称已被其他 MCP 服务器使用（全局或他人私有），请更换');
    }
  }

  private validateRequiredFields(server: McpServer): void {
    const type = server.serverType;
    if (type === TYPE_STDIO) {
      if (!hasText(server.command)) {
        throw new BusinessException(ErrorCode.PARAM_MISSING, 'STDIO 类型必须填写启动命令');
      }
      if (!hasText(server.argsJson) || server.argsJson === '[]') {
        throw new BusinessException(ErrorCode.PARAM_MISSING, 'STDIO 类型必须填写启动参数');
      }
    } else if (type === TYPE_HTTP) {
      if (!hasText(server.url)) {
        throw new BusinessException(ErrorCode.PARAM_MISSING, 'HTTP 类型必须填写服务器 URL');
      }
      if (!server.url!.startsWith('http://') && !server.url!.startsWith('https://')) {
        throw new BusinessException(ErrorCode.PARAM_INVALID, 'URL 必须以 http:// 或 https:// 开头');
      }
    } else {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '服务器类型必须为 STDIO 或 HTTP');
    }
  }

  private async fillOwnerNames(servers: McpServer[]): Promise<void> {
    if (!this.userLookup) return;
    const ownerIds = [...new Set(servers.map((s) => s.userId).filter((id): id is number => id != null && id !== GLOBAL_USER_ID))];
    if (ownerIds.length === 0) return;
    const nameMap = new Map<number, string>();
    for (const id of ownerIds) {
      const user = await this.userLookup.findById(id);
      if (user) {
        nameMap.set(id, hasText(user.displayName) ? user.displayName! : (user.username ?? ''));
      }
    }
    for (const s of servers) {
      if (s.userId != null && s.userId !== GLOBAL_USER_ID) s.userName = nameMap.get(s.userId) ?? null;
    }
  }
}
